### PR TITLE
[Codegen][GPU] Make GPUApplyTilingLevel run cleanup patterns

### PR DESCRIPTION
**Pull request by qedawkins**
This adds a few cleanup patterns to the end of `GPUApplyTilingLevel`, namely the pattern to merge consecutive insert/extract slices. The pipeline that is intended to use this pass is not sensitive to bufferization the same way and can freely run these patterns to simplify later loop hoisting patterns.

### CODE REVIEW

**Review by MaheshRavishankar — COMMENTED** [review]
Just a note, some of the insert/extract slice folding has been implicated in a race condition. See https://github.com/iree-org/iree/issues/17525#issuecomment-2142571780

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
The lit test seems like it isnt changing anything?

File: compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_apply_tiling_level.mlir [hunk 1]
```diff
@@ -106,3 +106,39 @@ module {
...
+
+// THREAD-LABEL: func.func @matmul_cleanup
+//       THREAD:   %[[B0:.+]] = hal.interface.binding.subspan set(0) binding(0)
+//       THREAD:   %[[B1:.+]] = hal.interface.binding.subspan set(0) binding(1)
+//       THREAD:   %[[A:.+]] = flow.dispatch.tensor.load %[[B0]]
+//       THREAD:   %[[B:.+]] = flow.dispatch.tensor.load %[[B1]]
+//       THREAD:   scf.for %{{.*}} = %c0 to %c64 step %c8
+//       THREAD:     scf.forall
+//   THREAD-DAG:       %[[LHS:.+]] = tensor.extract_slice %[[A]]
+//   THREAD-DAG:       %[[RHS:.+]] = tensor.extract_slice %[[B]]
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_apply_tiling_level.mlir:143]: The lit test is checking here that the `tensor.extract_slice` for the `for` tiling got fused with the slice for `forall` tiling

**Comment by qedawkins** [comment]
> Just a note, some of the insert/extract slice folding has been implicated in a race condition. See [#17525 (comment)](https://github.com/iree-org/iree/issues/17525#issuecomment-2142571780)

That is for folding insert/extract_slice ops into vector transfers. This is for merging `tensor.extract_slice(tensor.extract_slice)`

**Review by kuhar — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_apply_tiling_level.mlir (hunk 1 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_apply_tiling_level.mlir:143]: Ah, so it is generating the `scf.forall` too. Missed that.

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**